cmd/query/app/querysvc/query_service_v2.go

```diff
@@ -0,0 +1,218 @@
+type QueryServiceOptionsV2 struct {
+	ArchiveTraceReader tracestore.Reader
+	ArchiveTraceWriter tracestore.Writer
+	Adjuster           adjuster.Adjuster
+}
+
+// StorageCapabilities is a feature flag for query service
+type StorageCapabilities struct {
+	ArchiveStorage bool `json:"archiveStorage"`
+	// TODO: Maybe add metrics Storage here
+	// SupportRegex     bool
+	// SupportTagFilter bool
+}
+
+// QueryService contains span utils required by the query-service.
+type QueryServiceV2 struct {
+	traceReader      tracestore.Reader
+	dependencyReader depstore.Reader
+	options          QueryServiceOptionsV2
+}
+
+// GetTraceParams defines the parameters for retrieving traces using the GetTraces function.
+type GetTraceParams struct {
+	// TraceIDs is a slice of trace identifiers to fetch.
+	TraceIDs []tracestore.GetTraceParams
+	// RawTraces indicates whether to retrieve raw traces.
+	// If set to false, the traces will be adjusted using QueryServiceOptionsV2.Adjuster.
+	RawTraces bool
+}
+
+// TraceQueryParams represents the parameters for querying a batch of traces.
+type TraceQueryParams struct {
+	tracestore.TraceQueryParams
+	// RawTraces indicates whether to retrieve raw traces.
+	// If set to false, the traces will be adjusted using QueryServiceOptionsV2.Adjuster.
+	RawTraces bool
+}
+
+// NewQueryService returns a new QueryService.
+func NewQueryServiceV2(
+	traceReader tracestore.Reader,
+	dependencyReader depstore.Reader,
+	options QueryServiceOptionsV2,
+) *QueryServiceV2 {
+	qsvc := &QueryServiceV2{
+		traceReader:      traceReader,
+		dependencyReader: dependencyReader,
+		options:          options,
+	}
+
+	if qsvc.options.Adjuster == nil {
+		qsvc.options.Adjuster = adjuster.Sequence(
+			adjuster.StandardAdjusters(defaultMaxClockSkewAdjust)...)
+	}
+	return qsvc
+}
+
+// GetTraces retrieves traces with given trace IDs from the primary reader,
+// and if any of them are not found it then queries the archive reader.
+// The iterator is single-use: once consumed, it cannot be used again.
+func (qs QueryServiceV2) GetTraces(
+	ctx context.Context,
+	params GetTraceParams,
+) iter.Seq2[[]ptrace.Traces, error] {
+	getTracesIter := qs.traceReader.GetTraces(ctx, params.TraceIDs...)
+	// TODO: aggregate the traces
+	return func(yield func([]ptrace.Traces, error) bool) {
+		foundTraceIDs := make(map[pcommon.TraceID]struct{})
+		getTracesIter(func(traces []ptrace.Traces, err error) bool {
+			if err != nil {
+				return yield(nil, err)
+			}
+			for _, trace := range traces {
+				if !params.RawTraces {
+					qs.options.Adjuster.Adjust(trace)
+				}
+				resources := trace.ResourceSpans()
+				for i := 0; i < resources.Len(); i++ {
+					scopes := resources.At(i).ScopeSpans()
+					for j := 0; j < scopes.Len(); j++ {
+						spans := scopes.At(j).Spans()
+						for k := 0; k < spans.Len(); k++ {
+							span := spans.At(k)
+							foundTraceIDs[span.TraceID()] = struct{}{}
+						}
+					}
+				}
+			}
+			return yield(traces, nil)
```

Comment:
@yurishkuro We're currently yield here with all the traces that were found and then yielding again if there were some traces that weren't found in the primary storage but were found from the archive storage. Is this valid? Or do we only want to yield once?

Reply:
yes, it's valid